citrix_hypervisor/datadog_checks/citrix_hypervisor/check.py

```diff
@@ -0,0 +1,186 @@
+
+        self.tags = self.instance.get('tags', [])
+        self.xenserver = None  # type: Any
+        self._additional_tags = []  # type: List[str]
+
+        self.check_initializations.append(self._check_connection)
+
+    def _check_connection(self):
+        # type: () -> None
+        # Get the latest timestamp to reduce the length of the update endpoint response
+        r = self.http.get(self._base_url + '/host_rrd', params={'json': 'true'})
+        if r.status_code == 200:
+            self._last_timestamp = int(float(r.json()['lastupdate'])) - 60
+        else:
+            self.log.warning("Couldn't initialize the timestamp due to HTTP error %s, set it to 0", r.reason)
+
+    def _get_updated_metrics(self):
+        # type: () -> Dict
+        params = {
+            'start': self._last_timestamp,
+            'host': 'true',
+            'json': 'true',
+        }
+        r = self.http.get(self._base_url + '/rrd_updates', params=params)
+        r.raise_for_status()
+        # Response is not formatted for simplejson, it's missing double quotes " around the field names
+        # Explicitly use the python safe loader, the C binding is failing
+        # See https://github.com/yaml/pyyaml/issues/443
+        data = yaml.load(r.content, Loader=yaml.SafeLoader)
+
+        if data['meta'].get('end') is not None:
+            self._last_timestamp = int(data['meta']['end'])
+        else:
+            for key in data['meta'].keys():
+                if data['meta'][key] is None and key.startswith('end'):
+                    self._last_timestamp = int(key.split(':')[-1])
+
+        return data
+
+    def submit_metrics(self, data):
+        # type: (Dict) -> None
+        legends = data['meta']['legend']
+        values = data['data'][0]['values']
+        for i in range(len(legends)):
+            metric_name, tags = build_metric(legends[i], self.log)
+
+            if metric_name is not None:
+                self.gauge(metric_name, values[i], tags=self.tags + self._additional_tags + tags)
+
+    def _session_login(self, server):
+        # type: (Any) -> Dict[str, str]
+        # try:
+        session = server.session.login_with_password(
+            self.instance.get('username', ''), self.instance.get('password', '')
+        )
+        return session
+        # except Exception as e:
+        #     self.log.warning(e)
+        return {}
+
+    def _get_master_session(self, session):
+        # type: (Dict[str, str]) -> Dict[str, str]
+        # {'Status': 'Failure', 'ErrorDescription': ['HOST_IS_SLAVE', '192.168.101.102']}
+        if len(session.get('ErrorDescription', [])) < 1:
+            return {}
+
+        master_address = session['ErrorDescription'][1]
+        if not master_address.startswith('http://'):
+            master_address = 'http://' + master_address
+        master_xenserver = xmlrpclib.Server(master_address)
+
+        # Master credentials can be different, we can specify new `master_username` and
```

Comment:
Are these commonly used?

Reply:
What do you mean?

Reply:
Is it common for customers to use master credentials? If it's relatively common and simple to implement it might be worth implementing on the first release so come customers aren't blocked from using the integration.

Reply:
I don't know how common that is

Reply:
Ok, I think it's probably fine then if it isn't a well known thing